### PR TITLE
Akka 27103/streams zip all [#27103]

### PR DESCRIPTION
**Pull request by eyalfa**
## Purpose

This PR adds the zipAll functionality to akka streams Source and Flow interfaces. 
This operator is basically taken from scala's standard library, it zips two streams, right-padding the shorter side with default value provided by the caller.

please see #27103 for more details.

## References

References #27103

## Changes

* FlowOps now has zipAll and zipAllMat methods.
* javadsl Source and Flow also have these methods now, implemented based on the FlowOps methods.
* added documentation for the new operator.

## Background Context

This brings the akka streams API a bit closer to scala's standard library's collections.

### CODE REVIEW

**Comment by eyalfa** [comment]
cc: @johanandren

**Comment by johanandren** [comment]
OK TO TEST

**Comment by eyalfa** [comment]
@johanandren van u please review?

**Comment by johanandren** [comment]
Looks like the CI job is failing on formatting. See end of this section for details on running scalaFmt https://github.com/akka/akka/blob/master/CONTRIBUTING.md#sbt

We'll all be traveling this week, so please be patient with the review, we'll get to it. :)

**Comment by eyalfa** [comment]
@johanandren , missed the jenkins failure (looked at travis instead).
no worry, I'm a patient person :sunglasses:

**Comment by eyalfa** [comment]
@johanandren , still travelling? :sunglasses:

**Comment by johanandren** [comment]
Yup, back next week, I'll see if I have time to review tomorrow though.

File: akka-docs/src/main/paradox/stream/operators/Source-or-Flow/zipAll.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+@@@
+
+## Description
+
+Combines all elements from each of multiple sources into @scala[tuples] @java[*Pair*] and passes the @scala[tuples] @java[pairs] downstream.
+
+
+@@@div { .callout }
+
+**emits** when all of the inputs have an element available, once part (but not all) of the inputs complete their values are substituted by the provided defaults.
```
Comment (johanandren) [inline akka-docs/src/main/paradox/stream/operators/Source-or-Flow/zipAll.md]: I found this sentence a bit hard to parse, what about going for just  "when all of the inputs have an element available" and explaining about emitting the default elements in the description text instead?

**Review by johanandren — COMMENTED** [review]
Good work!

One thought is that both for Java and the case where one of the stream is completed there is an extra tuple allocation per stream pair, that is allocated and just thrown away. The Java case could be solved by a `zipAllWith` style operator (taking a function to construct the zipped object), but that would make this composition cleverness impossible and would require a custom graph stage. I can also not think of a way to avoid the tuple allocation for the case of one stream being completed without dropping down to a custom graph stage.

I think we can see that as a possible future improvement though and that this is good enough.

File: akka-docs/src/main/paradox/stream/operators/Source-or-Flow/zipAll.md (hunk 1 above)
Comment (eyalfa) [reply akka-docs/src/main/paradox/stream/operators/Source-or-Flow/zipAll.md]: Hi @johanandren , I think there's no escaping allocating the Tuple2 instances. actually the underlying zip operator behaves exactly the same (unless you consider the extra allocation n the map stage). in any case I think 'user space' operators are much simpler and easier to understand than custom stages.

looking at the code again I'd like to eliminate code duplication between zipAll and zipAllMat, I'll try to craft something and push a commit for that.

re. 'emits' documentation, what about:
at first emits when both inputs emit, and then as long as any input emits (coupled to the default value of the completed input).

in any case, this doc string has to be modified 6 (!) places (1 scala api, 4 java apis, documentation), right?

**Comment by eyalfa** [comment]
@johanandren , I've managed to refactor the zipAll logic into a flow and therefore eliminate code duplication between zipAll and zipAllMat.
I've also experienced with a way to eliminate the 'extra' tuple allocation, not sure if it makes sense or not, please have a look [here](https://github.com/akka/akka/commit/bfaedf4b1653ea3ba5f3f2d15db4726425600140).

I'd appreciate your advice regarding the documentation of this operator.

**Comment by eyalfa** [comment]
@johanandren , can u please have another look? thanks

**Comment by eyalfa** [comment]
@johanandren ping :sunglasses:

File: akka-docs/src/main/paradox/stream/operators/Source-or-Flow/zipAll.md (hunk 1 above)
Comment (johanandren) [reply akka-docs/src/main/paradox/stream/operators/Source-or-Flow/zipAll.md]: Yes that sounds like a good text to me.

**Review by johanandren — COMMENTED** [review]
I think this is good to merge with the emit doc updated.

The part about avoiding the alloc and providing a `zipAllWith` we can leave as a potential future improvement.

**Review by johanandren — APPROVED** [review]

**Comment by eyalfa** [comment]
@johanandren thanks for your approval, can you somehow request another committer to take a look? thanks

**Review by raboof — APPROVED** [review]
Good to have this mirroring the standard collection ops

**Comment by eyalfa** [comment]
@johanandren , @raboof thank you both for your review!
what's next? who's making the actual merge?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala [hunk 2]
```diff
@@ -2404,6 +2404,43 @@ trait FlowOps[+Out, +Mat] {
...
+  }
+
+  protected def zipAllFlow[U, A >: Out, Mat2](
+      that: Graph[SourceShape[U], Mat2],
+      thisElem: A,
+      thatElem: U): Flow[Out @uncheckedVariance, (A, U), Mat2] = {
+    case object passedEnd
+    val passedEndSrc = Source.repeat(passedEnd)
+    val left: Flow[Out, Any, NotUsed] = Flow[A].concat(passedEndSrc)
+    val right: Source[Any, Mat2] = Source.fromGraph(that).concat(passedEndSrc)
```
Comment (2m) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: Instead of cycling `passedEnd` element, here we could concat the flows with `Source.repeat(thisElem)` and `Source.repeat(thatElem)` respectively. Then the `map` operator below that converts `passedEnd` to `this/thatElem` would not be needed anymore.

Comment (eyalfa) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: In this case we will not be able to distinguish between an actual eof marker or a value from the original stream. It's quite possible that the provided default is also a valid value for stream elements.

Comment (raboof) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: I suspect the problem with that is when to complete the resulting stream: we might encounter `thisElem, thatElem` before both streams have reached their end.

Comment (2m) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala]: Ah yes. That makes sense!
